httpserver/handlers.go

```diff
@@ -1692,6 +1637,89 @@ func (s *Server) handleJobStderr(w http.ResponseWriter, r *http.Request, cluster
 	s.handleJobOutputFile(w, r, cluster, proc, "stderr", "Err")
 }
 
+// handleJobFile handles GET /api/v1/jobs/{cluster}.{proc}/files/{filename}
+func (s *Server) handleJobFile(w http.ResponseWriter, r *http.Request, cluster, proc int, filename string) {
+	if r.Method != http.MethodGet {
+		s.writeError(w, http.StatusMethodNotAllowed, "Method not allowed")
+		return
+	}
+
+	ctx := r.Context()
+
+	// Validate filename
+	if filename == "" {
+		s.writeError(w, http.StatusBadRequest, "Filename is required")
+		return
+	}
+
+	// Prevent path traversal attacks
+	if strings.Contains(filename, "..") || strings.Contains(filename, "/") || strings.Contains(filename, "\\") {
+		s.writeError(w, http.StatusBadRequest, "Invalid filename: path traversal not allowed")
+		return
+	}
+
+	// Build constraint for specific job
+	constraint := fmt.Sprintf("ClusterId == %d && ProcId == %d", cluster, proc)
+
+	// Download the job sandbox
+	sandboxCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	var sandboxBuf bytes.Buffer
+	errChan := s.schedd.ReceiveJobSandbox(sandboxCtx, constraint, &sandboxBuf)
```

Comment:
Here - and all the other `schedd.ReceiveJobSandbox` calls - switch to io.Pipe with a 8KB buffer.

DO NOT buffer the total file in memory.

Limit the amount of time you are writing to the HTTP response as you copy between the different buffers.

Refactor some of this logic into a standalone utility and use it in the other places you stream out the tarfile.